apis/v1alpha2/validation/httproute_test.go

```diff
@@ -572,6 +572,128 @@ func TestValidateHTTPPathMatch(t *testing.T) {
 	}
 }
 
+func TestValidateHTTPHeaderMatches(t *testing.T) {
+	tests := []struct {
+		name          string
+		headerMatches []gatewayv1a2.HTTPHeaderMatch
+		errCount      int
+	}{{
+		name:          "no header matches",
+		headerMatches: nil,
+		errCount:      0,
+	}, {
+		name: "no header matched more than once",
+		headerMatches: []gatewayv1a2.HTTPHeaderMatch{
+			{Name: "Header-Name-1", Value: "val-1"},
+			{Name: "Header-Name-2", Value: "val-2"},
+			{Name: "Header-Name-3", Value: "val-3"},
+		},
+		errCount: 0,
+	}, {
+		name: "header matched more than once (same case)",
+		headerMatches: []gatewayv1a2.HTTPHeaderMatch{
+			{Name: "Header-Name-1", Value: "val-1"},
+			{Name: "Header-Name-2", Value: "val-2"},
+			{Name: "Header-Name-1", Value: "val-3"},
+		},
+		errCount: 1,
+	}, {
+		name: "header matched more than once (different case)",
+		headerMatches: []gatewayv1a2.HTTPHeaderMatch{
+			{Name: "Header-Name-1", Value: "val-1"},
+			{Name: "Header-Name-2", Value: "val-2"},
+			{Name: "HEADER-NAME-2", Value: "val-3"},
+		},
+		errCount: 1,
+	}}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			route := gatewayv1a2.HTTPRoute{Spec: gatewayv1a2.HTTPRouteSpec{
+				Rules: []gatewayv1a2.HTTPRouteRule{{
+					Matches: []gatewayv1a2.HTTPRouteMatch{{
+						Headers: tc.headerMatches,
+					}},
+					BackendRefs: []gatewayv1a2.HTTPBackendRef{{
+						BackendRef: gatewayv1a2.BackendRef{
+							BackendObjectReference: gatewayv1a2.BackendObjectReference{
+								Name: gatewayv1a2.ObjectName("test"),
+								Port: utils.PortNumberPtr(8080),
+							},
+						},
+					}},
+				}},
+			}}
+
+			errs := ValidateHTTPRoute(&route)
+			if len(errs) != tc.errCount {
+				t.Errorf("got %d errors, want %d errors: %s", len(errs), tc.errCount, errs)
+			}
+		})
+	}
+}
```

Comment:
Perhaps a bit of a nitpick: How do you feel about verifying the error contents as well?

Reply:
Yeah I can do that; the rest of the tests in this file are only verifying error counts so I went with that for consistency but agree it'd be a bit better to verify specific errors.